File: bluetooth_scanner.rb
# bluetooth_scanner.rb — Ruby версия

require 'rble'
require 'json'
require 'csv'
require 'time'

class BluetoothScanner
  def initialize
    @devices = {}
    @running = true
  end

  def classify_rssi(rssi)
    if rssi >= -50
      ["Отлично", "\e[32m"]
    elsif rssi >= -70
      ["Средне", "\e[33m"]
    else
      ["Плохо", "\e[31m"]
    end
  end

  def print_table
    if @devices.empty?
      puts "\e[33m❌ Устройства не найдены.\e[0m"
      return
    end

    sorted = @devices.values.sort_by { |d| -d[:rssi] }

    puts "\n\e[36m📊 Найденные устройства:\e[0m"
    puts "┌" + "─" * 75 + "┐"
    puts "│ #{'Устройство'.ljust(25)} #{'MAC-адрес'.ljust(20)} #{'RSSI'.ljust(10)} #{'Статус'.ljust(15)} │"
    puts "├" + "─" * 75 + "┤"

    sorted.each do |d|
      name = d[:name] || "Неизвестно"
      name = name[0...22] + "..." if name.length > 25
      status, color = classify_rssi(d[:rssi])
      rssi_str = "#{d[:rssi]} dBm"
      puts "│ #{name.ljust(25)} #{d[:address].ljust(20)} #{color}#{rssi_str.ljust(10)}\e[0m #{color}#{status.ljust(15)}\e[0m │"
    end

    puts "└" + "─" * 75 + "┘"
  end

  def save_results
    data = {
      timestamp: Time.now.iso8601,
      devices: @devices.values
    }

    File.write("devices.json", JSON.pretty_generate(data))
    puts "\e[32m💾 Сохранено JSON: devices.json\e[0m"

    CSV.open("devices.csv", "w") do |csv|
      csv << ["Name", "Address", "RSSI"]
      @devices.values.each do |d|
        csv << [d[:name] || "Неизвестно", d[:address], d[:rssi]]
      end
    end
    puts "\e[32m💾 Сохранено CSV: devices.csv\e[0m"
  end

  def start
    puts "\e[36m📶 Bluetooth Scanner (RSSI) (Ruby)\e[0m"
    puts "🔍 Сканирование устройств... (нажмите Ctrl+C для остановки)"

    trap("INT") do
      @running = false
      puts "\n⏹️ Сканирование остановлено."
      print_table
      save_results
      exit
    end

    # Используем RBLE для сканирования
    begin
      RBLE.scan(timeout: 0) do |device|
        address = device.address
        name = device.name || "Неизвестно"
        rssi = device.rssi

        @devices[address] = {
          name: name,
          address: address,
          rssi: rssi
        }

        # Очищаем экран и выводим таблицу
        system("clear") || system("cls")
        print_table
      end
    rescue => e
      puts "\e[31m❌ Ошибка: #{e.message}\e[0m"
      puts "⚠️ Убедитесь, что Bluetooth включён и gem 'rble' установлен."
    end
  end
end

# Проверка зависимостей
begin
  require 'rble'
rescue LoadError
  puts "\e[31m❌ Установите rble: gem install rble\e[0m"
  exit 1
end

scanner = BluetoothScanner.new
scanner.start

File: bluetooth_scanner.php
<?php
// bluetooth_scanner.php — PHP версия

class BluetoothScanner {
    private $devices = [];
    private $running = true;

    public function __construct() {
        // Проверяем наличие bluetoothctl
        exec('which bluetoothctl', $output, $return);
        if ($return !== 0) {
            echo "\033[31m❌ bluetoothctl не найден. Установите bluez-utils.\033[0m\n";
            exit(1);
        }
    }

    private function classifyRSSI($rssi) {
        if ($rssi >= -50) {
            return ["Отлично", "\033[32m"];
        } elseif ($rssi >= -70) {
            return ["Средне", "\033[33m"];
        } else {
            return ["Плохо", "\033[31m"];
        }
    }

    private function scanDevices() {
        // Запускаем сканирование
        exec('bluetoothctl scan on 2>&1', $output, $return);
        
        // Читаем вывод в течение нескольких секунд
        $timeout = 10;
        $start = time();
        $devices = [];

        while (time() - $start < $timeout) {
            exec('bluetoothctl devices', $deviceOutput);
            foreach ($deviceOutput as $line) {
                if (preg_match('/Device\s+([0-9A-F:]+)\s+(.+)/', $line, $matches)) {
                    $address = $matches[1];
                    $name = $matches[2];
                    
                    // Получаем RSSI для устройства
                    exec("bluetoothctl info $address 2>&1", $infoOutput);
                    $rssi = -100;
                    foreach ($infoOutput as $infoLine) {
                        if (preg_match('/RSSI:\s+(-?\d+)/', $infoLine, $rssiMatch)) {
                            $rssi = (int)$rssiMatch[1];
                            break;
                        }
                    }
                    
                    $devices[$address] = [
                        'name' => $name,
                        'address' => $address,
                        'rssi' => $rssi
                    ];
                }
            }
            usleep(500000); // 0.5 сек
        }

        // Останавливаем сканирование
        exec('bluetoothctl scan off 2>&1');

        return $devices;
    }

    private function printTable($devices) {
        if (empty($devices)) {
            echo "\033[33m❌ Устройства не найдены.\033[0m\n";
            return;
        }

        // Сортировка по RSSI
        usort($devices, function($a, $b) {
            return $b['rssi'] - $a['rssi'];
        });

        echo "\n\033[36m📊 Найденные устройства:\033[0m\n";
        echo "┌" . str_repeat("─", 75) . "┐\n";
        echo "│ " . str_pad("Устройство", 25) . " " . str_pad("MAC-адрес", 20) . " " . str_pad("RSSI", 10) . " " . str_pad("Статус", 15) . " │\n";
        echo "├" . str_repeat("─", 75) . "┤\n";

        foreach ($devices as $d) {
            $name = $d['name'] ?: "Неизвестно";
            if (strlen($name) > 25) {
                $name = substr($name, 0, 22) . "...";
            }
            list($status, $color) = $this->classifyRSSI($d['rssi']);
            $rssiStr = $d['rssi'] . " dBm";
            echo "│ " . str_pad($name, 25) . " " . str_pad($d['address'], 20) . " " . $color . str_pad($rssiStr, 10) . "\033[0m " . $color . str_pad($status, 15) . "\033[0m │\n";
        }

        echo "└" . str_repeat("─", 75) . "┘\n";
    }

    private function saveResults($devices) {
        $data = [
            'timestamp' => date('c'),
            'devices' => array_values($devices)
        ];

        file_put_contents('devices.json', json_encode($data, JSON_PRETTY_PRINT | JSON_UNESCAPED_UNICODE));
        echo "\033[32m💾 Сохранено JSON: devices.json\033[0m\n";

        $csv = "Name,Address,RSSI\n";
        foreach ($devices as $d) {
            $csv .= ($d['name'] ?: "Неизвестно") . "," . $d['address'] . "," . $d['rssi'] . "\n";
        }
        file_put_contents('devices.csv', $csv);
        echo "\033[32m💾 Сохранено CSV: devices.csv\033[0m\n";
    }

    public function start() {
        echo "\033[36m📶 Bluetooth Scanner (RSSI) (PHP)\033[0m\n";
        echo "🔍 Сканирование устройств... (это займёт ~10 секунд)\n";

        $devices = $this->scanDevices();
        $this->printTable($devices);
        
        if (!empty($devices)) {
            $this->saveResults($devices);
        }
    }
}

// Обработка сигналов для корректного завершения
pcntl_signal(SIGINT, function() {
    echo "\n⏹️ Сканирование остановлено.\n";
    exit(0);
});

$scanner = new BluetoothScanner();
$scanner->start();
?>
